### DISCUSSION TITLE
Stata License

### DISCUSSION THREAD

**Discussion by julianmatthewman** [opensafely/documentation#299] [Q&A] [answered]
Hello,

I would like to know how to add my Stata license to run code with the stata-mp runner from the project.yaml on my machine. It isn't described in the documentation yet (https://docs.opensafely.org/actions-scripts/).

Many thanks in advance!

**Comment by bloodearnest** [comment, marked as answer]
Hi Julian

This isn't very well documented at the moment, but you can set an environment variable in your terminal, and then `opensafely run`  should then work fine with stata actions:

`export STATA_LICENSE="your license"`

However, there may be problems if your version of stata is not `stata-mp`, as that's the version OpenSAFELY uses, and I think the license has to match the same version. It's all a bit awkward! If so, let me know, and we'll need to figure out a way to solve the issue.

Thanks

**Reply by julianmatthewman** [reply]
Hi,

ah yes, it seems I don't have Stata MP but Stata IC. Any way to get it to work with that?

Thanks!

**Reply by bloodearnest** [reply]
Sadly, not easily.

We have a license for stata-mp, which is used in the protected backends, and can also be used in development by members of the `opensafely` github org. As you are collaborating on a single repo, this currently excludes you unfortunately.

I'll need to raise the issue internally, and see what's the best way to share our stata license with you.

**Reply by julianmatthewman** [reply]
I might be able to get my hands on a Stata MP license, but they consist of 3 separate codes (Serial number, Code and Authorization). How do I create the `STATA_LICENSE` to set as environment variable from these 3 codes?

**Reply by bloodearnest** [reply]
So, you'll need to use the 3 codes to register stata, which will produce a valid license. I think you can do this via the GUI, or via the `stinit` program that comes with stata. Once you've completed that process, you should have a `stata.lic` file in your stata installation. The contents of this file are what you need in STATA_LICENSE.

I have relayed your issue to the team and we discussing the best course of action. You are the first person using stata with OpenSAFELY that's not a member of the core OpenSAFELY team, so we've not had to solve this licensing for external users issue yet, and sadly stata's licensing model makes it tricky.

**Reply by remlapmot** [reply]
It might be helpful to note that the approach Code Ocean take in their Stata 16 dockerfile is to simply copy the `stata.lic` file (which has those 3 codes in on a single line separated by `!`, and then the user, org, and then I think the code for the Stata version) into the container. An example of their dockerfile is this capsule [here](https://doi.org/10.24433/CO.0587524.v1), see the lefthand pane and expand under environment (and also copied below). 
```
FROM registry.codeocean.com/codeocean/stata:16.0-ubuntu18.04

ARG DEBIAN_FRONTEND=noninteractive

COPY stata.lic /usr/local/stata/stata.lic

COPY postInstall /
RUN /postInstall
```

**Reply by julianmatthewman** [reply]
Thank you! Setting the license with `export STATA_LICENSE='your license'` worked. Note that one has to use `'` and not `"` for the license string as it contains exclamation marks(`!`) which serve a special function in bash.
Also for reference the Stata license is in the format Serial Number!Code!Authorization!User!Organisation!Version Code

**Reply by bloodearnest** [reply]
There's a technical issue here, and related legal one.

For the technical, we previously used to embed the license in the Docker, but this meant that we couldn't make that image public, which went against our transparency and audit policy, so now we pass the license details into the image when we run it via environment variable.

For the legal, we have a stata license for a limited number of users, which is used to run things on the secure backends. It's also used to run the same stata environment locally by the `opensafely` tool. Due to license limits we cannot currently increase the the number of users that use this license. We are working on a plan to fix this, but it may be a while.

**Reply by bloodearnest** [reply]
> Thank you! Setting the license with `export STATA_LICENSE='your license'` worked. Note that one has to use `'` and not `"` for the license string as it contains exclamation marks(`!`) which serve a special function in bash.
> Also for reference the Stata license is in the format Serial Number!Code!Authorization!User!Organisation!Version Code

Hah, yes, it can also include `$` characters for more bash fun. So single quotes are needed in bash. That was a head scratcher when we first encountered it.

Effectively, it's an opaque string as far as we are concerned, so the format doesn't matter too much.

I'm interesting in the last part though - you say it's a Version Code? I thought it was checksum or signature, as it seems to change every time we re-license.

**Reply by julianmatthewman** [reply]
Admittedly, I don't actually know what the last part of the license is, so Version Code might be wrong.

**Reply by sebbacon** [reply]
@julianmatthewman, could you confirm, was it an MP license, or for a different version? Thanks!

**Reply by sebbacon** [reply]
(for context: I'm updating the documentation. It'd also be handy to know how you located the license number in the correct format on _your_ machine)

**Reply by remlapmot** [reply]
The easiest way to find a Stata license number is just to open Stata - as it's in the output first printed to screen (Serial number line below)

```
  ___  ____  ____  ____  ____ ®
 /__    /   ____/   /   ____/      17.0
___/   /   /___/   /   /___/       MP—Parallel Edition

 Statistics and Data Science       Copyright 1985-2021 StataCorp LLC
                                   StataCorp
                                   4905 Lakeway Drive
                                   College Station, Texas 77845 USA
                                   800-STATA-PC        https://www.stata.com
                                   979-696-4600        stata@stata.com

Stata license: Unlimited-user 2-core network, expiring 21 Jan 2023
Serial number: 501709302483
```

For the rest of the info (Authorization and Code) as explained above - if you don't have the original documents all the info can be found in a text file, called `STATA.LIC` (on Windows) or `stata.lic` (macOS and Linux) which is usually at the top level of the folder of your Stata installation:

* On my Windows machine it's in `C:\Program Files\Stata17`
* On Linux it's again in the top level of my `stata17` folder which I put at the location Stata Corp. suggest `/usr/local/stata17/`
* On macOS it's in `/Applications/Stata/`

In the `stata.lic` file the different fields are separated by `!` characters. The license number is the first set of digits upto the first `!`.

**Reply by bloodearnest** [reply]
For the record, the STATA_LICENSE environment variable is passed through wholesale, so it needs to be the entire contents of your `stata.lic` file, not just the license serial number (which is not enough on its own for stata to accept).

**Reply by julianmatthewman** [reply]
> @julianmatthewman, could you confirm, was it an MP license, or for a different version? Thanks!

If I remember correctly, yes, it was an MP license.

**Comment by sebbacon** [comment]
Thanks all, I've updated the docs [here](https://docs.opensafely.org/actions-scripts/#stata)
